web/modules/custom/collation_fixer/src/CollationFixer.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\collation_fixer;
+
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Database\StatementInterface;
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Entity\Sql\SqlEntityStorageInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+
+/**
+ * Core class for fixing collations.
+ */
+final class CollationFixer {
+
+  /**
+   * Constructor.
+   */
+  public function __construct(
+    private readonly Connection $connection,
+    private readonly ModuleHandlerInterface $moduleHandler,
+    private readonly EntityTypeManagerInterface $entityTypeManager,
+    private readonly EntityFieldManagerInterface $fieldManager,
+  ) {}
+
+  /**
+   * Check table collations.
+   *
+   * @param string $table
+   *   Name of a table to check for collation correctness.
+   *   Defaults to all tables if not set.
+   *
+   * @return TableCollation[]
+   *   A list of tables with wrong collation or charset.
+   */
+  public function checkCollation(?string $table = NULL): array {
+    $schema = $this->getSchema($table);
+
+    $database_name = &drupal_static(__FUNCTION__);
+    if (empty($database_name)) {
+      $connection_options = $this->connection->getConnectionOptions();
+      $database_name = $connection_options['database'];
+    }
+
+    $fallback_charset = $this->getFallbackCharset();
+    $fallback_collation = $this->getFallbackCollation();
+
+    $wrong_collations = [];
+
+    foreach ($schema as $table_name) {
+      $schema_charset = $fallback_charset;
+      $charset_result = $this->connection->query(
+        'SELECT CCSA.character_set_name FROM information_schema.`TABLES` T,information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation AND T.table_schema = :table_schema AND T.table_name = :table_name',
+        [':table_schema' => $database_name, ':table_name' => $table_name]
+      );
+      $db_charset = ($charset_result instanceof StatementInterface) ? $charset_result->fetchField() : $fallback_charset;
+
+      $schema_collation = $fallback_collation;
+      $collation_result = $this->connection->query(
+        'SELECT TABLE_COLLATION FROM information_schema.tables WHERE TABLE_SCHEMA = :table_schema AND TABLE_NAME = :table_name',
+        [':table_schema' => $database_name, ':table_name' => $table_name]
+      );
+      $db_collation = ($collation_result instanceof StatementInterface) ? $collation_result->fetchField() : $fallback_collation;
+
+      if (($schema_charset != $db_charset) || ($schema_collation != $db_collation)) {
+        $wrong_collations[] = new TableCollation($table_name, $db_collation, $db_charset, $schema_collation, $schema_charset);
+      }
+    }
+
+    return $wrong_collations;
```

Comment:
I considered whether it would be a good idea to introduce an alter hook here.

This would allow users to specify expected collation/charset for specific tables instead of having to set it for the entire database.

Reply:
Good idea!

Reply:
I have tried to address this by adding an `collation_fixer_expected_collations` alter.